### PR TITLE
SNOW-1063719: [Local Testing] Add date_trunc support

### PR DESCRIPTION
**Pull request by sfc-gh-jrose**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #1063719

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   This change adds support for `date_trunc` to local testing. Most supported cases are handled by a call to pandas.floor, but several time units are not supported by pandas so equivalent logic is provided. I've also added related testing for date_trunc to validate correctness as well as type preservation.

### CODE REVIEW

File: src/snowflake/snowpark/functions.py [hunk 1]
```diff
@@ -4391,7 +4391,7 @@ def date_from_parts(
     return builtin("date_from_parts")(y_col, m_col, d_col)
 
 
-def date_trunc(part: ColumnOrName, expr: ColumnOrName) -> Column:
+def date_trunc(part: str, expr: ColumnOrName) -> Column:
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/functions.py:4394]: I agree this is the right change but I think this could be a BCR.

**Review by sfc-gh-yuwang — APPROVED** [review]
LGTM

File: src/snowflake/snowpark/functions.py [hunk 2]
```diff
@@ -4407,15 +4407,16 @@ def date_trunc(part: ColumnOrName, expr: ColumnOrName) -> Column:
...
         >>> df.select(date_trunc("HOUR", "a"), date_trunc("MINUTE", "a"), date_trunc("SECOND", "a")).collect()
-        [Row(DATE_TRUNC("HOUR", "A")=datetime.datetime(2020, 5, 1, 13, 0), DATE_TRUNC("MINUTE", "A")=datetime.datetime(2020, 5, 1, 13, 11), DATE_TRUNC("SECOND", "A")=datetime.datetime(2020, 5, 1, 13, 11, 20))]
+        [Row(DATE_TRUNC('HOUR', "A")=datetime.datetime(2020, 5, 1, 13, 0), DATE_TRUNC('MINUTE', "A")=datetime.datetime(2020, 5, 1, 13, 11), DATE_TRUNC('SECOND', "A")=datetime.datetime(2020, 5, 1, 13, 11, 20))]
         >>> df.select(date_trunc("QUARTER", "a")).collect()
-        [Row(DATE_TRUNC("QUARTER", "A")=datetime.datetime(2020, 4, 1, 0, 0))]
+        [Row(DATE_TRUNC('QUARTER', "A")=datetime.datetime(2020, 4, 1, 0, 0))]
     """
-    part_col = _to_col_if_str(part, "date_trunc")
+    if not isinstance(part, str):
+        raise ValueError("part must be a string")
```
Comment (sfc-gh-aalam) [inline src/snowflake/snowpark/functions.py:4417]: Would we get a SQL exception raised if a column name was passed instead of date part literal?

Comment (sfc-gh-jrose) [reply src/snowflake/snowpark/functions.py:4417]: Yes, if you try to pass an actual column instead of a valid date part it throws an exception.

**Pull request merged**